### ISSUE TITLE
:bug: Clearlogo and fanart laggy

### ISSUE THREAD

**Issue by HannesJo0139** [jurialmunkey/skin.arctic.fuse#1079]
### Skin section

Widgets

### Current Behavior

As can bee seen in the videos, it sometimes takes very long for the clearlogo to be updated. I am aware that it may take sometime initially, but then I would expect the Clearlogo of eg Vikings Valhalla to be cached. But as I navigate back to the Home Spotlight widget, it shows Vikings Valhalla with the Clearlogo of Oblivion. 

Furthermore, while recording I noticed that while the wrong clearlogo is shown, other fanart also is shown from the media the clearart actually belongs to. In the 2nd video you can see Artwork of Sense8 shown in the details view of Foundation. I then close the view and go up to spotlight where vikings is. The clearlogo and also the information dialog is populated with Sense8.

Until today I was using basic view and no spotlight at all. And I have had no issues. Just very rarely a singe Clearlogo taking some time. Today it started when I played a bit around and did a bit more complex setup including spotlights. Sometimes it is still working as expected, sometimes it is not. So far I have not been able to see any pattern. My last attempt to fix was to empty the TMDB Helper cache, but that actually made it even worse so I just decided to make videos and open a ticket.

https://github.com/user-attachments/assets/0fd96de1-aa28-4fce-91c0-ad025bcf9c9c

https://github.com/user-attachments/assets/9ad6d1e0-8640-4cd9-b5c9-2b74acb1b50d

### Expected Behavior

Fanart should always be in sync.

### Steps To Reproduce

Switch between different hubs, widgets and info screens.

### Checklist

- [X] I have made a separate issue for each bug that I am reporting.
- [X] I have searched the issues and this bug has not been reported.
- [X] I am using the most recent skin version.
- [X] I have confirmed that this bug does not occur in the default skin.

**Comment by HannesJo0139** [comment]
Some additional info and what else came into my mind:

- As soon as I disable Clearlogos, everything works fine. Then I also don't encounter additional fanart out-of-sync anymore. So the Clearlogo seems the root problem.
- I use Plex Kody Connect, what is basically the same as Emby plugin. I wonder if there maybe any interference.

**Comment by xyzfre** [comment]
https://github.com/user-attachments/assets/73eacd1a-1e95-4e83-9380-25020e9fe7eb

**Comment by HannesJo0139** [comment]
@xyzfre I see no issue in your vid what do you wanna show us?

**Comment by xyzfre** [comment]
My apologies last night got interrupted when I posted this video,,it was the show that everything works perfect and no lag. I can clear the cache and it still shows up that quick.

**Comment by HannesJo0139** [comment]
@jurialmunkey While it is not exactly specific about this local artwork issue I linked, I suspect it may indeed be a PKC related issue. As far as I understand I'm never going to have like Clearlogos in my Kodi library since PKC simply does not add it. Could it be that TMDb Helper this way always has to fetch a new Clearlogo again and again? Is there no Caching or sth like that? What do you think?

**Comment by jbelew** [comment]
> @jurialmunkey While it is not exactly specific about this local artwork issue I linked, I suspect it may indeed be a PKC related issue. As far as I understand I'm never going to have like Clearlogos in my Kodi library since PKC simply does not add it. Could it be that TMDb Helper this way always has to fetch a new Clearlogo again and again? Is there no Caching or sth like that? What do you think?
> 
> Just to be clear, I'm not trying to do anything with local artwork. There is no local Artwork anywhere, not in Plex and also not in Kodi. But it still looks like there is no TMDb helper caching.
> 
> Edit: I just found the cache folder and it is in fact full of cached Clearlogos. Maybe I'm gonna have a look into Kodi db.
> 
> Edit2: Some updates
> 
> * Checking my Kodi library via sqlite db browser, there is only art of the type fanart, posters and thumb available. There are no entries of clearlogo, as I expected. So I wonder why there would be any interference between PKC and TMDb Helper.
> * Also, I checked the TMDb Helper cache folder. Kodi is turned off. In the cache, there is a clearart logo of the show Foundation (It is in my continue watching widgets..). I start kodi, open the continue watching witget, focus Foundation. It takes about 20 seconds to load the clearart, even though it is definitely cached!
> * In the [Kodi log](https://paste.kodi.tv/gamolaliqe.kodi) I cannot find anything special. There is just some info about 404 errors when TMDb Helper tried to open [this](https://api.themoviedb.org/3/tv/93740/season/17?&append_to_response=credits,images,release_dates,content_ratings,external_ids,movie_credits,tv_credits,keywords,reviews,videos,watch/providers&region=DE&language=de-DE&include_image_language=de,null,en&include_video_language=de,null,en).
> 
> As this seems to end up being a TMDb helper issue, should we switch to an issue over there @jurialmunkey?

PKC has no impact on the performance of Clearlogos or Fanart on my system. I have a bit of a "custom" install where PKC manages just TV Shows and Music, then use the native Kodi library for Movies making it easy to compare.

Clearlogo's aren't fast in either case, even after caching. It takes about 500ms to appear after the the fanart is updated on my oDroid N2+ regardless of the DB source, but it's not unreasonable.

**Comment by HannesJo0139** [comment]
Thanks for your reply, that is interesting. I just did a total re-setup of Kodi and PKC, and it got somewhat better than in the first videos, but still takes commonly about 10…20 secs to load a clearlogo that is already cached. But contrary to your experience, it is not only after watching sth. I also see the issue right after startup.

**Comment by jurialmunkey** [comment]
I think jbelew is in the right direction here. This looks like some sort of Kodi lock up either blocking disk io (ie to read cache) or service thread itself. 

It reminds me of a scenario that sometimes happens where Kodi gets stuck loading one plugin thread and because it never completes it then blocks all other widgets/containers  loading next in the queue (it's not that but feels similar). 

There's always going to be some delay in the service because it has to wait for focus to change before it runs its code. But not this long, and especially not if the ratings and art manipulations are already cached. Retrieving from cache is not that complicated or processor intensive.

PKC shouldn't really have any real impact. The only thing I can think might be remotely related is if you have a very large library, then you probably should turn off TMDbhelper kodi db lookups in tmdbhelper settings. And especially if PKC is writing the database often. I can't remember if PKC is the one that writes to kodi library directly or if it's the other plex plugin but if PKC is writing to the library directly rather than just access via an addon path then that could be the issue. 

If kodi db is very large and changes frequently then tmdbhelper has to do a lot of processing to pull the details (and every change to the library it then needs to redo again). Long shot but disabling the setting for getting additional details from kodi db might be worthwhile in this case (make sure to hit okay to save setting and will need to restart kodi for it to come into effect for the service).

**Comment by xyzfre** [comment]
> @xyzfre Are you using PKC? Windows, Linux?

Windows 11

**Comment by HannesJo0139** [comment]
@jurialmunkey thanks for your reply, PKC is the one managing the Kodi db. But I would not consider my lib very large. Roughly 600 movies and 150 tv shows (ok it is not small either). However, the tmdb helper setting to catch additional details from kodi lib db is already turned off. So it is propably unrelated to PKC.

So since the log does not show anything special I really have no clue how to debug this. Any ideas anyone? @jbelew do you know if there are any other issues or support requests for this already? I was unable to find anything so far. Maybe I should open a thread in the Kodi forum.

**Comment by jbelew** [comment]
> @jurialmunkey thanks for your reply, PKC is the one managing the Kodi db. But I would not consider my lib very large. Roughly 600 movies and 150 tv shows (ok it is not small either). However, the tmdb helper setting to catch additional details from kodi lib db is already turned off. So it is propably unrelated to PKC.
> 
> So since the log does not show anything special I really have no clue how to debug this. Any ideas anyone? @jbelew do you know if there are any other issues or support requests for this already? I was unable to find anything so far. Maybe I should open a thread in the Kodi forum.

I've looked, but not found anything. I honestly think that's a bit of a rabbit hole since I'm running CoreElec which includes an optimized, under development OS stack (down to the kernel) as part of the install, which none of the other reporters on this issue are using. My hunch on the resource contention I'm seeing is it's NFS and the crappy DNS-323's that I'm still using to host some of my library; locking up the file system while waiting for a "Wake on LAN" response. 

What I do know though is that up to date versions of the PKC and TMDB Helper perform at comprarible speeds vs. the native library when installed "properly".  This is with an extremely large library of 6000+ episodes and 30000+ music files. As I stated previously, movies are handled natively by Kodi. 

Since the reporter is on Debian, you might want to backup your ".kodi" directory and try a fresh install with just TMDB Helper and see if the issue persists. Get it configured and see if the issue can be reproduced, then introduce PKC.

**Comment by Rinicko100** [comment]
> https://github.com/user-attachments/assets/73eacd1a-1e95-4e83-9380-25020e9fe7eb

The kind of speed I dream about... how is this even possible. What device do you use? also are the fanart and clear logos from your local library or from TMDB helper?

**Comment by HannesJo0139** [comment]
Ok here is a first impression. On another PC I have a freshly installed EndavourOS, freshly installed Kodi with Arctic Fuse. No PKC, nothing else. All the movies and shows are directly added via smb share. So far, I see no i/o blocking like issue. Clearart loading looks okay. What I see again is this mixing-in of fanart that actually belongs to one of the previously selected elements. It looks like the skin would re-use templates (eg. information view) that still have previous content loaded, and while loading the new content you can actually see the stuff from before. It seems there are two types of fanart. First type is stuff coming directly from Kodi db(?). This type is always loaded extremely fast. 2nd type seems to be stuff Arctic Fuse loads from tmdb helper(?). This takes a bit longer.

I also tried some other skins: Estuary, Copacetic, Titan Bingie Mod, Arctic Horizon 2. None of them shows this behavior. Where clearart is still loading I see text instead, where other fanart is still loading there is a placeholder.

I'll do some more testing over the weekend to see if I can also find a trigger of this i/o blocking, but for now @jurialmunkey could it be Arctic Fuse is currently not using placeholders properly? (And in the same manner not using text as clearart placeholder as long as new clearart is still loading)

https://github.com/user-attachments/assets/ae70956c-4467-4007-b168-c31104f59b0c

**Comment by HannesJo0139** [comment]
Now I've introduced PKC and it got worse. Almost as bad as in my initial videos. My first impression is that PKC is just doing some additional background tasks that are not problematic per se but just seem to make that tmdb helper race condition even worse. 

https://github.com/user-attachments/assets/a7c699fd-9b4f-4198-a7d8-51ebcf02ec52


So I went into the PKC settings and disabled some stuff that looks like 'could do things in the background' (background library updates and stuff like that). In the beginning it felt like it got a bit better again but in the end I'm not that certain anymore.

https://github.com/user-attachments/assets/998de37f-77d3-4b9f-9509-a53877079154


Disabling PKC feels way better (almost the same like before I setup PKC), but still not satisfying. 

https://github.com/user-attachments/assets/c8fca71b-8133-481a-8e05-bac8ab8db6d8


I think it still is kind of uncertain but I would say propably PKC is just making an existing issue in arctic fuse / tmdb helper more obvious. In the end of the day I can see that most elements like plot texts, background fanart, etc is working perfectly well while Clearart and mixing-in of incorrect fanart is just a symptom of that tmdb helper race condition. I know I cannot expect everything to be super fast, especially when navigating very fast. But I wonder if there may be any possibility arctic fuse could kind of skip incoming tmdb helper results that are not up-to-date anymore. @jurialmunkey would be happy to hear your thoughts :)

**Comment by jbelew** [comment]
Are you waiting for PKC to do it's initial startup sync before you compare? If you don't, both the TMDB and PKC python processes default to uncapped threads counts.

**Comment by HannesJo0139** [comment]
Yepp I waited. 

..Turning back to the htpc. I was just using it these days normally and ignoring the laggy art. Now I just noticed that the issue has become quite rare here atm. I can see almost no "i/o blocks" (Lets hope it's not coming back lol). Looking into the tmdb helper cache there is almost 1000 cropped clearart logos. So a large percentage seems to be cached now. What comes into my mind: Could it be that simply tmdb helper is blocking itself, as long as many art was not cached yet?

**Comment by jurialmunkey** [comment]
If the art is not cached, you have to wait. When you focus an uncached item, tmdbhelper has to download the logo and fanart (either from disk location for local, or online for http), then apply the crop/blur transformations, then save the transformed file to the cache folder.

Once it does that, it writes the location of the cached file to a window property. The window property is what points to the location of the image file in the texture control in the skin. 

Kodi sees the file location changed in the property and thus the texture control image path, so it must then cache that new image into its system texture cache before it can render it in the control. 

If the image is uncached, kodi will continue to display the previous image in the texture control until the new image is available in the loader cache to render (or if the image path changes again).

Tmdbhelper does the artwork and ratings manipulation in a separate thread from its main service. So, broadly speaking, if you see the ratings update (the online ones like trakt or rottentomatoes etc) but there is still a lag in the artwork, then this lag is likely due to Kodi background loader delays in caching the new artwork. However, if you see that the ratings are also delayed, then this is more likely a delay in tmdbhelpers thread for artwork/ratings.

If the delay is mostly kodi side, then nothing can be done.

If the delay is tmdbhelper side, there's not really much that can be done either - the time it takes to download, manipulate, and save the new image is however long it takes.

One way you might reduce load is by disabling tmdbhelper service readaheads (off top of my head the option is in the expert setting section). The readaheads try to precache items ahead of the current one while service is sitting idle - so it's a bit of a mixed bag performance wise to disable it. Disabling gets more resources for the main thread as the readaheads are never running in the background but obviously a precached item will display faster than one which is not. 

Really the only way to reduce the load for tmdbhelper here is by disabling blur and/or crop effects in the skin by changing background from real to fake blur (or even fanart) so that it doesn't need to do the blur transformation, or turning off logos so that it doesn't need to do the crop. That's sort of sidestepping the issue though as a big question is why is io reading of the artwork is so slow in the first place.

**Comment by HannesJo0139** [comment]
Alright, thanks for the info. I absolutely understand there are delays we’d have to accept. I was more like thinking in the direction that we may try do prevent Arctic Fuse from showing outdated info. I think this is the actual issue here. Like making sure on focus change (or a short time after) to show no ratings until the correct ones are available. Making sure to show text instead of wrong Clearlogo and prevent wrong artwork from mixing in.

I obviously have no idea of kodis skin system but in general I would say incoming info has to be (in)validated. Do you think this would be possible somehow?

**Comment by xyzfre** [comment]
What device are you using? Never seen Arctic Fuse act this slow,, I have tried using it on a Nvidia Shiey and my mini PC. It works amazing.

**Comment by HannesJo0139** [comment]
Htpc is a Intel i3 9100. 16G RAM. SSD. The following tests on a PC with Ryzen 3800x, 32G RAM, SSD.

Are U using network locations, Linux, Windows?

**Comment by HannesJo0139** [comment]
@jbelew Are you using PKC with DirectPath (if so, SMB, NFS?) or default Addon Paths?

**Comment by jbelew** [comment]
> @jbelew Are you using PKC with DirectPath (if so, SMB, NFS?) or default Addon Paths?

Direct paths over NFS. Haven't we settled on no impact between PKC and TMDB Helper or do I need to create a video to prove it?

**Comment by HannesJo0139** [comment]
@jbelew exactly, the actual point behind my question was to know if you use SMB as I do 😉

**Comment by HannesJo0139** [comment]
LOL so finally I gave Windows a try. Even uncached on the very first try everything is MUCH faster than ever before using different Linuxes. Did anyone say stuff like uncached cropped clearest has to be a bit slow? Not here…

**Comment by HannesJo0139** [comment]
Looks like Kodi would not guarantee that external scripts can load stuff in time, sadly
https://github.com/xbmc/xbmc/issues/25766#issuecomment-2381933348
